### PR TITLE
[MS-ABI] skip generate comdat for vftable defined with internal alias.

### PR DESCRIPTION
**Pull request by jyu2-git**
We got a error:
`LLVM ERROR: Associative COMDAT symbol '??_7?$T@V<lambda_0>@@@@6B@' is not a key for its COMDAT`

Current we create internal alias for vftable when lambd is used.
For the test, IR generate:
```
 $"??_7?$T@V<lambda_0>@@$0A@@@6b@" = comdat any

@0 = private unnamed_addr constant { [2 x ptr] } { [2 x ptr] [ptr @"??_R4?$T@V<lambda_0>@@$0A@@@6b@", ptr @"?c@b@@UEAAXXZ"] }, comdat($"??_7?$T@V<lambda_0>@@$0A@@@6b@")

@"??_7?$T@V<lambda_0>@@$0A@@@6b@" = internal unnamed_addr alias ptr, getelementptr inbounds ({ [2 x ptr] }, ptr @0, i32 0, i32 0, i32 1)
```

According LLVM language reference manual section on COMDATs:
There are some restrictions on the properties of the global object. It,
or an alias to it, must have the same name as the COMDAT group when
targeting COFF. The contents and size of this object may be used during
link-time to determine which COMDAT groups get selected depending on the
selection kind. Because the name of the object must match the name of the
COMDAT group, the linkage of the global object must not be local; local
symbols can get renamed if a collision occurs in the symbol table.

So one way to fix this is to not create comdat for the alias.

@0 = private unnamed_addr constant { [2 x ptr] } { [2 x ptr] [ptr @"??_R4?$T@V<lambda_0>@@@@6B@", ptr @"?c@?$T@V<lambda_0>@@@@UEAAXXZ"] }

### CODE REVIEW

**Comment by efriedma-quic** [comment]
Why are we creating a comdat here in the first place?  Creating a comdat associated with an internal symbol doesn't do anything useful: an internal symbol can't be defined in any other translation unit, so the comdat will never be discarded.

If we allow creating a comdat associated with an internal symbol, the backend should handle it correctly; it's a bug if it crashes like you've described.  (I'm not sure how to reproduce the crash with your testcase, though.)

**Comment by efriedma-quic** [comment]
> Why are we creating a comdat here in the first place? 

I guess that's b2615aa4... but I'm pretty sure we fixed -fdata-sections so the backend automatically generates comdats.  So maybe we should just revert that.

> If we allow creating a comdat associated with an internal symbol, the backend should handle it correctly; it's a bug if it crashes like you've described. (I'm not sure how to reproduce the crash with your testcase, though.)

LangRef says "Because the name of the object must match the name of the COMDAT group, the linkage of the global object must not be local; local symbols can get renamed if a collision occurs in the symbol table."  I think that means the IR clang generates is illegal, but I guess nothing in the backend actually checks this.

**Comment by jyu2-git** [comment]
Hi @efriedma-quic,

Thanks for looking into this.  First for test case:
We are having difficult to reproduce the problem with small test case, the error come from very big application, which related various heuristics to trigger in a specific way to get the symbol renaming to happen, then for the inliner to occur on selected instances, and then for dead function elimination to happen...

Second, 
>>LangRef says "Because the name of the object must match the name of the COMDAT group, the linkage of the global object must not be local; local symbols can get renamed if a collision occurs in the symbol table." I think that means the IR clang generates is illegal, but I guess nothing in the backend actually checks this.

Except that error we see, there is no other verification check that.

Third:
>>I guess that's https://github.com/llvm/llvm-project/commit/b2615aa44d0cabd821b7afe5acdc847af9937c76... but I'm pretty sure we fixed -fdata-sections so the backend automatically generates comdats. So maybe we should just revert that.

Let me try to revert that, to see if I can get through all tests, just worry about ABI change.

Thank you so much for looking into.

Jennifer

**Comment by rnk** [comment]
> I guess that's https://github.com/llvm/llvm-project/commit/b2615aa44d0cabd821b7afe5acdc847af9937c76... but I'm pretty sure we fixed -fdata-sections so the backend automatically generates comdats. So maybe we should just revert that.

That seems reasonable, however, there are other QOI reasons for attempting to generate globally unique names for lambdas. The debug info which describes them relies on having unique mangled names, if there are duplicates, the debugger will not be able to find the correct type information.

I think our strategy to ensure internal lambdas like this are globally unique is that they appear in some externally visible scope which can only be defined once, and that scope gets mangled into the name of the lambda. Is that happening here? I don't have an MS demangler handy to confirm against the test case.

**Comment by efriedma-quic** [comment]
> That seems reasonable, however, there are other QOI reasons for attempting to generate globally unique names for lambdas

Sure.  But we wouldn't want to special-case vftables for that.

> I think our strategy to ensure internal lambdas like this are globally unique is that they appear in some externally visible scope which can only be defined once, and that scope gets mangled into the name of the lambda. Is that happening here? I don't have an MS demangler handy to confirm against the test case.

In the testcase, the lambda is just at global scope.  Not sure what you're expecting to happen here.

**Comment by rnk** [comment]
> In the testcase, the lambda is just at global scope. Not sure what you're expecting to happen here.

I think that's what I'm trying to get at, this strategy is incomplete and didn't work.

I think a principled fix to the general problem would be to incorporate the anonymous namespace hash into the names of all internal lambdas. The point where we check the lambda mangling number is the point where we know if we have an internal or external lambda and decide accordingly:
https://github.com/llvm/llvm-project/blob/11f52f783a64acef671a2b1648a40ba683ad4717/clang/lib/AST/MicrosoftMangle.cpp#L1125

Also this logic has been duplicated, that doesn't seem good:
https://github.com/llvm/llvm-project/blob/11f52f783a64acef671a2b1648a40ba683ad4717/clang/lib/AST/MicrosoftMangle.cpp#L268

To your point, we shouldn't special case vftables, since this solves a more general problem of debug info quality.

**Comment by jyu2-git** [comment]
Thanks @efriedma-quic and @rnk, both ideas should fix our problem.

I am choosing @efriedma-quic's idea, since this has less impact for
compile.

I also tried @rnk's idea, I don't fell comfort by adding
getAnonymousNamespaceHash’s string
for all lambda name mangling.

Thanks all!

**Review by efriedma-quic — APPROVED** [review]
I see unrelated changes on the branch?  Otherwise LGTM

**Comment by jyu2-git** [comment]
> I see unrelated changes on the branch? Otherwise LGTM

This my first time using this new review.  I need merge them in.  I revert my first change add new one.  But it seems it is not merged.

**Comment by efriedma-quic** [comment]
The simplest thing to do is rebase and force-push.  You can alternatively try to use a merge commit, but I'm not completely sure how github would react to that.

**Comment by efriedma-quic** [comment]
Looks fine now.  (You might want to change the title before merging.)

**Review by rnk — APPROVED** [review]

File: clang/test/CodeGenCXX/type-metadata.cpp [hunk 1]
```diff
@@ -125,9 +125,9 @@
 // MS-TYPEMETADATA: comdat($"??_7B@@6B0@@"), !type [[B8:![0-9]+]]
 // MS-TYPEMETADATA: comdat($"??_7B@@6BA@@@"), !type [[A8]]
 // MS-TYPEMETADATA: comdat($"??_7C@@6B@"), !type [[A8]]
-// MS-TYPEMETADATA: comdat($"??_7D@?A0x{{[^@]*}}@@6BB@@@"), !type [[B8]], !type [[D8:![0-9]+]]
-// MS-TYPEMETADATA: comdat($"??_7D@?A0x{{[^@]*}}@@6BA@@@"), !type [[A8]]
-// MS-TYPEMETADATA: comdat($"??_7FA@?1??foo@@YAXXZ@6B@"), !type [[A8]], !type [[FA8:![0-9]+]]
+// MS-TYPEMETADATA: private unnamed_addr constant { [3 x ptr] } { [3 x ptr] [ptr @"??_R4D@?A0xFA4729A9@@6BB@@@", ptr @"?g@B@@UEAAXXZ", ptr @"?h@D@?A0xFA4729A9@@UEAAXXZ"] }, !type [[B8]], !type [[D8:![0-9]+]]
```
Comment (jyu2-git) [inline clang/test/CodeGenCXX/type-metadata.cpp:128]: The problem is AnonymousNamespaceHas, this need to change to {{.*}}.

**Pull request merged**